File: src/backend/core/keyManager.ts
import { Mutex } from 'async-mutex';

export class KeyManager {
  private keys: string[];
  private index: number;
  private mutex: Mutex;

  constructor(keys: string[]) {
    this.keys = keys;
    this.index = 0;
    this.mutex = new Mutex();
  }

  async getNextKey(): Promise<string | null> {
    const release = await this.mutex.acquire();
    try {
      if (this.keys.length === 0) return null;
      const key = this.keys[this.index];
      this.index = (this.index + 1) % this.keys.length;
      return key;
    } finally {
      release();
    }
  }

  async markInvalid(badKey: string): Promise<void> {
    const release = await this.mutex.acquire();
    try {
      this.keys = this.keys.filter(k => k !== badKey);
      if (this.keys.length === 0) this.index = 0;
    } finally {
      release();
    }
  }
}

File: backend/app/core/key_manager.py
import threading

class KeyManager:
    def __init__(self, keys):
        self.keys = keys
        self.index = 0
        self._lock = threading.Lock()

    def get_next_key(self):
        with self._lock:
            if not self.keys:
                return None
            key = self.keys[self.index]
            self.index = (self.index + 1) % len(self.keys)
            return key

    def mark_invalid(self, bad_key):
        with self._lock:
            self.keys = [k for k in self.keys if k != bad_key]
            if not self.keys:
                self.index = 0
